bpf/lib/audit.h

```diff
@@ -0,0 +1,85 @@
+/*
+ *  Copyright (C) 2016-2019 Authors of Cilium
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+/*
+ * Policy audit notification via perf event ring buffer
+ *
+ * API:
+ * int send_audit_notify(skb, src, dst, dst_id, reason, direction)
+ *
+ * If AUDIT_NOTIFY is not defined, the API will be compiled in as a NOP.
+ */
+
+#ifndef __LIB_AUDIT__
+#define __LIB_AUDIT__
+
+#include "dbg.h"
+#include "drop.h"
+#include "events.h"
+#include "common.h"
+#include "utils.h"
+#include "metrics.h"
+
+#ifdef AUDIT_NOTIFY
+
+/**
+ * send_audit_notify
+ * @skb:	socket buffer
+ * @src:	source identity
+ * @dst:	destination identity
+ * @dst_id:	designated destination endpoint ID
+ * @reason:	reason for drop
+ * @direction:	packet direction
+ *
+ * Generate a notification to indicate a packet was allowed due to the audit mode.
+ */
+static inline void send_audit_notify(struct __sk_buff *skb, __u32 src, __u32 dst,
+				     __u32 dst_id, int reason, __u8 direction)
+{
+	update_metrics(skb->len, direction, -reason);
+
+	uint64_t skb_len = (uint64_t)skb->len, cap_len = min((uint64_t)TRACE_PAYLOAD_LEN, (uint64_t)skb_len);
+	uint32_t hash = get_hash_recalc(skb);
+	struct drop_notify msg = {
+		.type = CILIUM_NOTIFY_AUDIT,
+		.subtype = -reason,
+		.source = EVENT_SOURCE,
+		.hash = hash,
+		.len_orig = skb_len,
+		.len_cap = cap_len,
+		.version = NOTIFY_CAPTURE_VER,
+		.src_label = src,
+		.dst_label = dst,
+		.dst_id = dst_id,
+		.unused = 0,
+	};
+	skb_event_output(skb, &EVENTS_MAP,
+			 (cap_len << 32) | BPF_F_CURRENT_CPU,
+			 &msg, sizeof(msg));
+}
+
+#else
+
+static inline void send_audit_notify(struct __sk_buff *skb, __u32 src, __u32 dst,
+				     __u32 dst_id, int reason, __u8 direction)
+{
+	update_metrics(skb->len, direction, -reason);
```

Comment:
I wonder if we should add one extra bit to `struct metrics_key` to indicate when the drop is actually an audit-drop rather than a real policy drop?

Reply:
@joestringer I'm split on this one, it feels like both metrics will have similar semantics: count packet stats per decision and logic for each decision is exactly the same. Audit mode is a global flag so I think there shouldn't be confusion in terms of metrics. I can change it if you think it will be less ambiguous to have these metrics separate.

Reply:
I wouldn't worry about it too much right now unless someone else feels strongly; issue #9580 has a comment suggesting a more granular audit mode which I think could provide stronger reasoning for this, but at the same time we could always just split it out if and when that idea is implemented.

Reply:
@joestringer I initially tried to implement granular annotation based audit mode but had an issue with that. Policies whitelist traffic instead of blacklisting so technically there is no direct connection between dropped packet and a policy since packet is dropped because it's not affected by any policy.  So when we decided to drop a packet there is no way to link it back to a policy with audit annotation.

Reply:
True, if there are N policies then there are N potential reasons for dropping the packet (as they all put it in default-deny) so the CNP-style annotation is a bit questionable. There is an item for 1.8 which may introduce blocklist policies (https://github.com/cilium/cilium/issues/7111). I was also considering this from the perspective of per-pod annotations.